### ISSUE TITLE
The type or namespace name 'YamlDotNet' could not be found

### ISSUE THREAD

**Issue by Matbtz** [aaubry/YamlDotNet#501]
Hello, 

Sorry if this has already been discussed but I could not find any topic on this.
I installed the plugin via the Unity asset store and installed it in my projet Asset/Plugin.

I then created the following class :

*using UnityEngine;
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections;
using System.Collections.Generic;
**using YamlDotNet.Serialization;**
**using YamlDotNet.Serialization.NamingConventions;**

public class Datastore
{

    public class ClassList
    {
        //public List<Class> classes { get; set; }
    }

    public Datastore()
    {
    }

    public void Load()

    {
        LoadClasses();
    }

    private void LoadClasses()
    {
        TextAsset classesYaml = Resources.Load("Classes") as TextAsset;

        StringReader input = new StringReader(classesYaml.text);

        Deserializer deserializer = new Deserializer(namingConvention: new CamelCaseNamingConvention());

        ClassList classList = deserializer.Deserialize<ClassList>(input);

        //Class.AddItems(classList.classes);
    }

}*

The problem that I encounter is the following for both YamlDotNet imports: 
Assets\Resources\Datastore.cs(8,7): error CS0246: The type or namespace name 'YamlDotNet' could not be found (are you missing a using directive or an assembly reference?) 

Any idea on how to solve this ?

Thanks a lot

**Comment by Matbtz** [comment]
I continued to make some tests and it seems that the package does not install well or is not recognized.
Indeed when I install it via the asset store, the package is not present in the package manager.

I tried to install it manually via git url but unity could not install it and finished with an error...

I also tried with NuGet installer in Visual Studio, the package was installed. I thought it worked for a minute but when back in unity editor I still had the “The type or namespace name 'YamlDotNet' could not be found”.

I tried to delete the package in my project folder and reimport several times but could not succeed to make it work yet...

I’ll continue to do more tests in the coming days.

**Comment by aaubry** [comment]
I do not distribute YamlDotNet on the asset store. You should contact the maintainer of that distribution, or you could use the YamlDotNet NuGet package, which contains a Unity version of the library.

**Comment by aaubry** [comment]
Sorry, I was replying to the first message, but I've now noticed that you tried with the NuGet package as well. I've never used Unity so I don't know why that's not working. Do you have some code that you could share so that I can attempt to reproduce the problem ?

**Comment by aaubry** [comment]
I tried installing Unity but I don't know what to do with it. I don't even see where I can edit code or add a NuGet package. If you provide some instructions, I can try to help.

**Comment by Matbtz** [comment]
Ok that's nice !
You can  : 

- create a New projet in File > New Project (2D for example)
- Then on the Project window (left down of the screen normally) you can right click on anywhere in Asset > Create >C# script and name it Database
- Double click on it to open it with Visual Studio (it comes with Unity I think)
- In Visual Studio you can go in Tools > NuGet Package manager console and import the package with Install-Package YamlDotNet
- Then you can copy paste this class in the database file as a test, I found this class from a tutorial on how to use YamlDotNet (https://www.bendangelo.me/2016/02/21/creating-a-rpg-database-in-unity/): 

```
using UnityEngine;
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

public class Database {

    public class ClassList {
        //public List<Class> classes { get; set; }
    }

    public Database() {
    }

    public void Load() {
        LoadClasses();
    }

    private void LoadClasses() {
        TextAsset classesYaml = Resources.Load("Classes") as TextAsset;

        StringReader input = new StringReader(classesYaml.text);

        Deserializer deserializer = new Deserializer(namingConvention: new CamelCaseNamingConvention());

        ClassList classList = deserializer.Deserialize<ClassList>(input);

        //Class.AddItems(classList.classes);
    }

}
```

From there I have two issues, on Visual Studio the 
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
seems ok but when going back into Unity, in the console, it tells me that the " **The type or namespace name 'YamlDotNet' could not be found**" for both namespace.

And also the : Deserializer deserializer = new Deserializer(namingConvention: new CamelCaseNamingConvention());
is not working as the Deserializer does not have a method with parameters, should I just remove the parameter ?

By installing the package with NuGet, it installed in the project folder in Packages but does not seem to be recognized by Unity.

I am currently trying to drag and drop manually this package into unity on the Project tree but did not succeed to make it work yet . 
Instead I encounter more issue : _Multiple precompiled assemblies with the same name YamlDotNet.dll included or the current platform. Only one assembly with the same name is allowed per platform. Assembly paths:_ 

If you manage to reproduce and find something I would appreciate it but don't worry it's not a huge issue for me!

**Comment by aaubry** [comment]
I tried those steps but got stuck at:
> In Visual Studio you can go in Tools > NuGet Package manager console and import the package with Install-Package YamlDotNet
The package manager console complains that I don't have any project loaded, which is true. The only thing that is open in Visual Studio is the C# file itself. It is not inside a project.

I've checked [this page](https://stackoverflow.com/questions/53447595/nuget-packages-in-unity), which mentions ways to use NuGet packages, but it doesn't seem straightforward.

**Comment by Matbtz** [comment]
Hmm ok that's weird, it seems that when opening the C# script from Unity, visual Studio detects the project and select by default the Assembly-CSharp file that is in the folder root
![image](https://user-images.githubusercontent.com/66409283/84770093-42193880-afd7-11ea-9d8f-1f3c976f7730.png)

Ok I will check how they are using it in the page you linked

**Comment by aaubry** [comment]
Well I already had Visual Studio installed. Maybe the Unity installer would have installed it with some extension, I don't know.

Anyway, I managed to make it build by using the following steps:
- Open a command prompt.
- Install YamlDotNet:  
  ```
   nuget install YamlDotNet
  ```  
  Alternatively you could download the package from nuget.org as explained in that thread.
- Copy the file `YamlDotNet.dll` from `YamlDotNet.8.1.2\lib\net35-client` to `Assets\Plugins` (had to create the Plugins directory).

The code that you provided needs to be updated by replacing:
```c#
Deserializer deserializer = new Deserializer(namingConvention: new CamelCaseNamingConvention());
```
by:
```c#
var deserializer = new DeserializerBuilder()
    .WithNamingConvention(CamelCaseNamingConvention.Instance)
    .Build();
```
It now builds, but I have no way of knowing whether it actually works.

**Comment by Matbtz** [comment]
Ah yes yesterday I tried to take only one dll file but could not figure out which one !
Anyway I followed the steps given in the thread you gave me, I installed NuGet asset on Unity and it imported YamlDotNet just fine. And with the correction you just mentionned for the deserializer I have no more errors :)

I'll be able to test if everything works fine tonight but I think you succeeded to unlock the situation for me.

Thank you so much :)

**Comment by Matbtz** [comment]
It is indeed working well, thank you very much for your help :)

**Comment by fer-gamboa** [comment]
I know this issue is closed, but I thought of contributing my solution in case somebody else needs help. I tried the steps here but couldn't get YamlDotNet to work with Unity. At the end, I found this [Unity package](https://github.com/GlitchEnzo/NuGetForUnity) that directly connects to NuGet, and was able to quickly install YamlDotNet in Unity (2019.3.1f1) without any issues. No need to copy dlls back and forth or anything else. Quite straightforward.
